core/drivers/rtc/rtc.c

```diff
@@ -5,10 +5,113 @@
 
 #include <assert.h>
 #include <drivers/rtc.h>
-#include <tee_api_types.h>
+
+/* Time fields shift values to fit date in a 64bit word */
+#define RTC_CMP_YEAR_SHIFT	U(36)
+#define RTC_CMP_MONTH_SHIFT	U(32)
+#define RTC_CMP_DAY_SHIFT	U(27)
+#define RTC_CMP_HOUR_SHIFT	U(22)
+#define RTC_CMP_MINUTES_SHIFT	U(16)
+#define RTC_CMP_SECONDS_SHIFT	U(10)
 
 struct rtc *rtc_device;
 
+static const uint8_t rtc_months_days[] = {
+	31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31
+};
+
+static unsigned long long rtc_time_to_ms(struct optee_rtc_time *time)
+{
+	uint64_t time_ms = 0;
+	unsigned int n = 0;
+	uint64_t days = 0;
+	uint64_t year = 0;
+
+	if (!time->tm_mday || time->tm_year < rtc_device->range_min.tm_year ||
+	    time->tm_year > rtc_device->range_max.tm_year)
+		panic();
+
+	/* Elapsed milliseconds since midnight */
+	time_ms = time->tm_ms;
+	time_ms += time->tm_sec * MS_PER_SEC;
+	time_ms += time->tm_min * MS_PER_MIN;
+	time_ms += time->tm_hour * MS_PER_HOUR;
+
+	/* Elapsed days in current month and in past months of current year */
+	days = time->tm_mday - 1;
+	for (n = 0; n < time->tm_mon; n++)
+		days += rtc_get_month_days(n, time->tm_year);
+
+	/* Elapsed days pasted years */
+	year = time->tm_year;
+	days += (year * 365) + (year / 4) - (year / 100) + (year / 400);
+
+	/* Accumulate the elapsed days */
+	time_ms += days * MS_PER_DAY;
+
+	return time_ms;
+}
+
+bool rtc_is_a_leap_year(uint32_t year)
+{
+	return !(year % 4) && ((year % 100) || !(year % 400));
+}
+
+uint8_t rtc_get_month_days(uint32_t month, uint32_t year)
+{
+	if (rtc_is_a_leap_year(year) && month == 1)
+		return rtc_months_days[month] + 1;
+	else
+		return rtc_months_days[month];
+}
+
+int rtc_timecmp(struct optee_rtc_time *a, struct optee_rtc_time *b)
+{
+	uint64_t tm_a = 0;
+	uint64_t tm_b = 0;
+
+	tm_a = SHIFT_U64(a->tm_year, RTC_CMP_YEAR_SHIFT) +
+	       SHIFT_U64(a->tm_mon, RTC_CMP_MONTH_SHIFT) +
+	       SHIFT_U64(a->tm_mday, RTC_CMP_DAY_SHIFT) +
+	       SHIFT_U64(a->tm_hour, RTC_CMP_HOUR_SHIFT) +
+	       SHIFT_U64(a->tm_min, RTC_CMP_MINUTES_SHIFT) +
+	       SHIFT_U64(a->tm_sec, RTC_CMP_SECONDS_SHIFT) +
+	       a->tm_ms;
+	tm_b = SHIFT_U64(b->tm_year, RTC_CMP_YEAR_SHIFT) +
+	       SHIFT_U64(b->tm_mon, RTC_CMP_MONTH_SHIFT) +
+	       SHIFT_U64(b->tm_mday, RTC_CMP_DAY_SHIFT) +
+	       SHIFT_U64(b->tm_hour, RTC_CMP_HOUR_SHIFT) +
+	       SHIFT_U64(b->tm_min, RTC_CMP_MINUTES_SHIFT) +
+	       SHIFT_U64(b->tm_sec, RTC_CMP_SECONDS_SHIFT) +
+	       b->tm_ms;
+
+	return CMP_TRILEAN(tm_a, tm_b);
+}
+
+unsigned long long rtc_diff_calendar_ms(struct optee_rtc_time *ref1,
+					struct optee_rtc_time *ref2)
+{
+	if (rtc_timecmp(ref1, ref2) < 0)
+		panic();
```

Comment:
Sorry to insist. As this is a generic function, there is no real need to support only positive differences.
Let's return a signed value and for consistency (even is unlikely) use `SUB_OVERFLOW()` to return `LLONG_MIN`/`LLONG_MAX` upon diff overflows.

This allow you to test the full time value against range min and range max in commit "drivers: stm32_rtc: check data consistency in rtc_set_time()".

Reply:
Ok, let me come up with something with signed values